### PR TITLE
HPCC-26275 Add ws_resources method for listing Apps/Websites

### PR DESCRIPTION
**Pull request by wangkx**
The method returns the information about web sites associated
with an HPCC cloud environment. ECLWatch will use the information
to link the web sites.

Signed-off-by: wangkx kevin.wang@lexisnexis.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by GordonSmith — COMMENTED** [review]
@kwang - can you add suitable helm attributes for the Roxie Test page (the one on port 8002) along with this PR so I can test a real world example?

File: esp/services/ws_resources/ws_resourcesService.cpp [hunk 1]
```diff
@@ -31,4 +31,76 @@ bool CWsResourcesEx::onServiceQuery(IEspContext& context, IEspServiceQueryReques
...
+            {
+                StringBuffer annotationName;
+                annotationsItr->query().getName(annotationName);
+                if (!hasPrefix(annotationName, "hpcc.eclwatch.io_f", false))
+                    continue;
+
+                //Add all hpcc.eclwatch.io annotations.
+                Owned<IEspNamedValue> annotation = createNamedValue();
+                annotation->setValue(annotations->queryProp(annotationName));
+                annotationName.replaceString("_f", "/");
```
Comment (rpastrana) [inline esp/services/ws_resources/ws_resourcesService.cpp:94]: what introduces the "_f" ?

File: helm/hpcc/values.yaml [hunk 2]
```diff
@@ -437,9 +437,16 @@ esp:
     ## Annotations can be specified on a service - for example to specify provider-specific information such as service.beta.kubernetes.io/azure-load-balancer-internal-subnet
     #annotations:
     #  service.beta.kubernetes.io/azure-load-balancer-internal-subnet: "mysubnet"
+    #  hpcc.eclwatch.io/enabled: "true"
+    #  hpcc.eclwatch.io/description: "some description"
     ## You can also specify labels on a service
     #labels:
     #  mylabel: "3"
+    ## Links specify the web links for a service. The web links may be shown on ECLWatch.
```
Comment (rpastrana) [inline helm/hpcc/values.yaml:445]: These links are static (likely external to k8s cluster) and should only be acknowledged in service:esp/application:eclwatch/links

File: esp/scm/ws_resources.ecm [hunk 3]
```diff
@@ -34,9 +51,20 @@ ESPresponse [nil_remove, exceptions_inline] ServiceQueryResponse
...
+ESPresponse [nil_remove, exceptions_inline] WebAppQueryResponse
+{
+    ESParray<ESPstruct ECLWatchIOLink> ECLWatchIOLinks;
+    ESParray<ESPstruct ServiceLink> ServiceLinks;
+};
+
+ESPservice [auth_feature("ResourceQueryAccess:ACCESS"), noforms, version("1.01"), exceptions_inline("./smc_xslt/exceptions.xslt")] WsResources
 {
     ESPmethod [auth_feature("ResourceQueryAccess:READ")] ServiceQuery(ServiceQueryRequest, ServiceQueryResponse);
+    ESPmethod [auth_feature("ResourceQueryAccess:READ"), min_ver("1.01")] WebAppQuery(WebAppQueryRequest, WebAppQueryResponse);
```
Comment (rpastrana) [inline esp/scm/ws_resources.ecm:67]: EclWatchLinkQuery?

**Review by rpastrana — CHANGES_REQUESTED** [review]
@wangkx code looks good, but a few comments.

File: esp/scm/ws_resources.ecm [hunk 4]
```diff
@@ -34,9 +51,20 @@ ESPresponse [nil_remove, exceptions_inline] ServiceQueryResponse
...
 
-ESPservice [auth_feature("ResourceQueryAccess:ACCESS"), noforms, version("1.00"), exceptions_inline("./smc_xslt/exceptions.xslt")] WsResources
+ESPrequest WebAppQueryRequest
+{
+};
+
+ESPresponse [nil_remove, exceptions_inline] WebAppQueryResponse
+{
+    ESParray<ESPstruct ECLWatchIOLink> ECLWatchIOLinks;
+    ESParray<ESPstruct ServiceLink> ServiceLinks;
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply esp/scm/ws_resources.ecm:61]: @GordonSmith as the likely only consumer, what are your thoughts on this layout?

[the remark this answers is not in the document]
Comment (wangkx) [reply esp/scm/ws_resources.ecm:61]: The ECLWatchIOLink and ServiceLink have different contents. They can be merged into one. The client can read them from different locations based on the source. @GordonSmith  do you want me merge them into one?

File: esp/scm/ws_resources.ecm [hunk 5]
```diff
@@ -23,6 +25,21 @@ ESPStruct HPCCService
     bool TLSSecure;
 };
 
+ESPStruct ECLWatchIOLink
+{
+    string Name;
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/scm/ws_resources.ecm:30]: The Name in ECLWatchIOLink should be changed to ServiceName. The Name in ServiceLink is the name of a web link.

**Comment by wangkx** [comment]
@GordonSmith I did 2 annotation tests. 
1. Install a sample service based on the steps in https://artifacthub.io/packages/helm/cloudecho/hello#tl-dr. Then, I added the annotations using the following commands:
kubectl annotate service my-hello hpcc.eclwatch.io/enabled="true"
kubectl annotate service my-hello hpcc.eclwatch.io/description="A hello-world program"
kubectl annotate service my-hello hpcc.eclwatch.io/protocol="http"
kubectl annotate service my-hello hpcc.eclwatch.io/hostname="localhost"
2. In helm/hpcc/values.yaml, add the following lines under the line of the 'servicePort: 8002':
annotations:
  hpcc.eclwatch.io/enabled: "true"
  hpcc.eclwatch.io/description: "Roxie Test page"
  hpcc.eclwatch.io/protocol="http"
  hpcc.eclwatch.io/hostname="localhost"

File: esp/services/ws_resources/ws_resourcesService.cpp [hunk 6]
```diff
@@ -31,4 +31,76 @@ bool CWsResourcesEx::onServiceQuery(IEspContext& context, IEspServiceQueryReques
         FORWARDEXCEPTION(context, e,  ECLWATCH_INTERNAL_ERROR);
     }
     return false;
+}
+
+//1. Get Web App links defined in ComponentConfig() service/links.
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/services/ws_resources/ws_resourcesService.cpp:36]: Please see my comment https://github.com/hpcc-systems/HPCC-Platform/pull/15395#discussion_r733778736

File: helm/hpcc/values.yaml (hunk 2 above)
Comment (wangkx) [reply helm/hpcc/values.yaml:470]: Will add comment per offline discussion.

**Comment by wangkx** [comment]
@rpastrana please review my changes.

**Review by rpastrana — CHANGES_REQUESTED** [review]
@wangkx looks good. Left a comment regarding wording describing the service.annotations

**Review by rpastrana — APPROVED** [review]
@wangkx

File: esp/services/ws_resources/ws_resourcesService.cpp [hunk 7]
```diff
@@ -31,4 +31,76 @@ bool CWsResourcesEx::onServiceQuery(IEspContext& context, IEspServiceQueryReques
...
+           throw makeStringExceptionV(ECLWATCH_INTERNAL_ERROR, "The '%s' returns empty response.", command.str());
+
+        IArrayOf<IConstECLWatchIOLink>& eclWatchIOLinks = resp.getECLWatchIOLinks();
+        Owned<IPropertyTree> outputTree = createPTreeFromJSONString(output.str());
+        Owned<IPropertyTreeIterator> outputItemItr = outputTree->getElements("items");
+        ForEach(*outputItemItr)
+        {
+            IPropertyTree& outputItemTree = outputItemItr->query();
+            IPropertyTree* annotations = outputItemTree.queryPropTree("metadata/annotations");
+            if (!annotations || !annotations->getPropBool("hpcc.eclwatch.io_fenabled", false))
```
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/services/ws_resources/ws_resourcesService.cpp:74]: So it's kind of interesting, we have made the service agnostic about which application is running, but the meta-data we look for is still application specific.  (hpcc.eclwatch.io/enabled).  I wonder if we should 1. make the metadata not mention eclwatch (hpcc.io/enabled), or 2. use the currently running "esp application" to build the meta string to look for.  ```VStringBuffer tag("hpcc.%s.io_fenabled", espcfg->queryProp("@application"));```

The "esp application" for eclwatch ```(espcfg->queryProp("@application"))``` is "eclwatch", so the user wouldn't notice the difference, but other applications could then use their own names.

@wangkx @rpastrana Any thoughts?

File: esp/services/ws_resources/ws_resourcesService.cpp [hunk 8]
```diff
@@ -31,4 +31,77 @@ bool CWsResourcesEx::onServiceQuery(IEspContext& context, IEspServiceQueryReques
...
+            discoveredWebLinks.append(*discoveredWebLink.getLink());
+        }
+#endif
+    }
+    catch(IException* e)
+    {
+        FORWARDEXCEPTION(context, e,  ECLWATCH_INTERNAL_ERROR);
+    }
+    return false;
 }
```
Comment (afishbeck) [inline esp/services/ws_resources/ws_resourcesService.cpp:107]: Needs newline.

**Review by afishbeck — COMMENTED** [review]
@wangkx One question, and one very minor formatting issue.

**Comment by wangkx** [comment]
> I wonder if we should 1. make the metadata not mention eclwatch (hpcc.io/enabled), or 2. use the currently running "esp application" to build the meta string to look for. VStringBuffer tag("hpcc.%s.io_fenabled", espcfg->queryProp("@application"));

I like the #2. The hpcc.io/enabled is too general. @rpastrana your opinion?

**Comment by rpastrana** [comment]
> 
> 
> > I wonder if we should 1. make the metadata not mention eclwatch (hpcc.io/enabled), or 2. use the currently running "esp application" to build the meta string to look for. VStringBuffer tag("hpcc.%s.io_fenabled", espcfg->queryProp("@Application"));
> 
> I like the #2. The hpcc.io/enabled is too general. @rpastrana your opinion?

Based on the current implementation, I prefer #2 as well.

**Comment by wangkx** [comment]
@afishbeck I just squashed the commits. Please do a review.

**Review by afishbeck — APPROVED** [review]
@wangkx looks good.

**Pull request merged**